### PR TITLE
RHIROS #448 - Make IsConfigured API response consistent with Openapi spec

### PR DESCRIPTION
**Pull request by apuntamb**
In addition to maintaining consistency between the API response and the Openapi spec, also refactored queries inside the `is_configured` api to only count those systems which have a performance profile similar to the query for `systems` api.

Also, did a minor refactoring to define common/default queries required by all APIs in one method.

EDIT - Removed the refactoring bit and would instead handle that change in a separate PR.

### CODE REVIEW

File: ros/api/v1/hosts.py [hunk 1]
```diff
@@ -25,30 +25,46 @@
...
+        system_query = default_queries(ident)
 
+        last_reported = (
+            db.session.query(PerformanceProfile.system_id, func.max(PerformanceProfile.report_date).label('max_date')
+                             )
+            .filter(PerformanceProfile.system_id.in_(system_query.subquery()))
+            .group_by(PerformanceProfile.system_id)
+            .subquery()
+        )
+        print(last_reported)
```
Comment (saltgen) [inline ros/api/v1/hosts.py:48]: Shall we use LOG.info, in case we need this one?

File: ros/api/v1/hosts.py [hunk 2]
```diff
@@ -25,30 +25,46 @@
...
-            }
+        system_query = default_queries(ident)
 
+        last_reported = (
+            db.session.query(PerformanceProfile.system_id, func.max(PerformanceProfile.report_date).label('max_date')
+                             )
+            .filter(PerformanceProfile.system_id.in_(system_query.subquery()))
+            .group_by(PerformanceProfile.system_id)
+            .subquery()
+        )
```
Comment (saltgen) [inline ros/api/v1/hosts.py:47]: :ok_hand:

File: ros/api/v1/hosts.py (hunk 1 above)
Comment (apuntamb) [inline ros/api/v1/hosts.py:48]: @saltgen need to remove the print statement. Forgot to do so. :)

File: ros/api/v1/hosts.py [hunk 3]
```diff
@@ -25,30 +25,46 @@
 ]
 
 
+def default_queries(ident, additional_filter=None):
+    account_query = db.session.query(RhAccount.id).filter(RhAccount.account == ident['account_number']).subquery()
+    if additional_filter is None:
+        return db.session.query(System.id).filter(System.account_id.in_(account_query))
+    return db.session.query(System.id).filter(System.account_id.in_(account_query)).filter(additional_filter)
+
```
Comment (saltgen) [inline ros/api/v1/hosts.py:33]: Thanks for the modularization :raised_hands: 
Looks like only the `account_number` is required from `ident`, shall we pass just the account_number
in such a case?
Also, I think we can move the function to utils.py or some other file that would be appropriate

File: ros/api/v1/hosts.py (hunk 1 above)
Comment (saltgen) [reply ros/api/v1/hosts.py:48]: Np. I donno why I deleted the related comment :rofl: Oh well

**Comment by saltgen** [comment]
Done with the comments for now. Thanks for the changes! :smiley:

File: ros/api/v1/hosts.py (hunk 3 above)
Comment (apuntamb) [reply ros/api/v1/hosts.py:33]: @saltgen You mean directly passing `ident['account_number']` instead of `ident`? Because if so, in any case we have to fetch the account number from ident, so I don't foresee much of a difference.
Do let me know if that sounds okay or not?

**Review by PreetiW — APPROVED** [review]
@apuntamb Thank you for making the changes, I have tested this PR locally and it is working fine. Also checked it against Dashboard code and it won't break anything there

image

image

File: ros/api/v1/hosts.py (hunk 3 above)
Comment (saltgen) [reply ros/api/v1/hosts.py]: Agreed, there is not much real-time gain in this, but in terms of readability, IMO this looks clearer.

```
def default_queries(account_number: str, additional_filter: bool = None):
    ...
    ...
```

P.S. Added the type hints just to convey what I really mean, not required to add them.

Comment (apuntamb) [reply ros/api/v1/hosts.py]: Done.

**Review by PreetiW — CHANGES_REQUESTED** [review]
Hello @apuntamb 
I was checking filters functionality and the API is giving 500 due to attached screenshot
image

can you please check the reason behind this?

**Review by PreetiW — APPROVED** [review]
Thank you looking at the issue @apuntamb , checked with the latest commit and filtering is working fine. I also checked systems details page and it is working as ecpected 💯 

image

image

**Comment by apuntamb** [comment]
@saltgen Any additional comments or can we merge this?

File: ros/api/v1/hosts.py [hunk 4]
```diff
@@ -27,28 +29,34 @@
...
+        account_number = identity(request)['identity']['account_number']
+        system_query = default_queries(account_number)
 
-        if system_count <= 0:
-            return {
-                'success': False,
-                'code': 'NOSYSTEMS',
-                'count': system_count
-            }
+        last_reported = (
```
Comment (kgaikwad) [inline ros/api/v1/hosts.py:35]: I would like to understand why did you add this extra query?

File: ros/api/v1/hosts.py [hunk 5]
```diff
@@ -27,28 +29,34 @@
...
+            db.session.query(PerformanceProfile.system_id, func.max(PerformanceProfile.report_date).label('max_date')
+                             )
+            .filter(PerformanceProfile.system_id.in_(system_query.subquery()))
+            .group_by(PerformanceProfile.system_id)
+            .subquery()
+        )
+        query = (
+            db.session.query(PerformanceProfile, System, RhAccount)
```
Comment (kgaikwad) [inline ros/api/v1/hosts.py:43]: Any specific reason you are loading all columns from all three tables and that is to get counts?

**Review by kgaikwad — CHANGES_REQUESTED** [review]

**Comment by saltgen** [comment]
> @saltgen Any additional comments or can we merge this?

Nope, lgtm. Thanks @apuntamb! :raised_hands:

File: ros/api/v1/hosts.py (hunk 4 above)
Comment (apuntamb) [reply ros/api/v1/hosts.py:35]: @kgaikwad IMO, inside the /systems api, systems are listed based on Performance Profile filter i.e. we show systems only which have Performance Profile associated with the respective system.
It would make more sense to use the same query inside /is_configured api to keep the system counts consistent in both api responses.
Thoughts?

**Comment by saltgen** [comment]
@apuntamb On second thought could you add an assert comparison in the test `test_is_configured`, between
count from systems endpoint and is_configured?

File: ros/api/v1/hosts.py [hunk 6]
```diff
@@ -30,25 +30,32 @@ def get(self):
...
-        system_count = db.session.query(System.id)\
-            .filter(System.account_id.in_(account_query)).count()
-        systems_with_suggestions = db.session.query(System.id)\
-            .filter(System.account_id.in_(account_query))\
-            .filter(System.number_of_recommendations > 0).count()
-        systems_waiting_for_data = db.session.query(System.id)\
-            .filter(System.account_id.in_(account_query))\
-            .filter(System.state == 'Waiting for data').count()
+        system_query = db.session.query(System.id).filter(System.account_id.in_(account_query))
+        last_reported = (
```
Comment (kgaikwad) [inline ros/api/v1/hosts.py:34]: @apuntamb  - Could you please add new issue for checking PerformanceProfile records?
Original issue only talks about inconsistent response formats.

File: tests/test_api_endpoints.py [hunk 7]
```diff
@@ -38,7 +38,7 @@ def test_status():
         assert response.json["status"] == "Application is running!"
 
 
-def test_is_configured(auth_token, db_setup, db_create_account, db_create_system):
+def test_is_configured(auth_token, db_setup, db_create_account, db_create_system, db_create_performance_profile):
```
Comment (kgaikwad) [inline tests/test_api_endpoints.py:41]: @apuntamb  - why `db_create_performance_profile` required here?

Comment (apuntamb) [reply tests/test_api_endpoints.py]: fixed.

**Comment by apuntamb** [comment]
@kgaikwad As discussed over call with @pgarciaq , will handle **not** showing all systems i.e. **only** those systems which are shown through `/systems` api response in `is_configured` api too, in a separate PR.

**Review by kgaikwad — APPROVED** [review]
LGTM :+1: 
As discussed, refactoring and mismatched system count issue should be handled separately.
